Support alibaba cloud tablestore

xumingming opened the pull request:
Test plan - (Please fill in how you tested your changes)

Tested on my desktop with real config.

```
== RELEASE NOTES ==

General Changes

*  Add a new connector: tablestore for Alibaba Cloud's Tablestore service

beinan commented on presto-tablestore/pom.xml:
```diff
@@ -0,0 +1,238 @@
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
```
indentation style looks not following other sections

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,251 @@
+{
+    enum Type
+    {
+        auto,
```
I think we prefer using upper case for enum

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,251 @@
+                    .collect(Collectors.toSet());
+            return IndexFirstSet.custom(tables);
+        }
+        String str = "use hint like these: " +
```
thinking if we could extract 'str' as an constant 
can we have a meaningful name rather than 'str'?

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,251 @@
+    private static SchemaTableName checkAndAssemblySchemaTableName(Optional<String> currentConnectionSchema,
+            String hintKey,
+            String unit)
+            throws IllegalArgumentException
```
hmmm, why we need explicitly throws a runtime exception in a private method?

xumingming replied:
You mean the exception should be throw in a public method?

beinan replied:
Not really, I'm just curious 
1. Why we need a "throws" for a RuntimeException?  We don't have to, am I right? (IllegalArgumentException is a runtime/unchecked exception, isn't it?)
2. In case you just want to use 'throws' to make a strong contract to the caller, but 'checkAndAssemblySchemaTableName' is a private method.  I cannot see the benefits of the 'throws' here.

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,99 @@
+
+    public static Field findField(Class<?> clazz, String name, Class<?> type)
+    {
+        checkArgument(clazz != null, "Class must not be null");
```
we might prefer to use requireNonNull instead of checkArgument here

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,99 @@
+            }
+            searchType = searchType.getSuperclass();
+        }
+        return null;
```
make this method returns an Optioanl<Field> ?  your call

zhenxiao replied:
+1. let's make it Optional, instead of returning null

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,99 @@
+    public static void rethrowRuntimeException(Throwable ex)
+    {
+        if (ex instanceof RuntimeException) {
+            throw (RuntimeException) ex;
```
unnecessary cast

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,99 @@
+
+    public static void handleInvocationTargetException(InvocationTargetException ex)
+    {
+        rethrowRuntimeException(ex.getTargetException());
```
why we need a separate method here? can we inline rethrowRuntimeException?

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,99 @@
+            return field.get(target);
+        }
+        catch (IllegalAccessException ex) {
+            handleReflectionException(ex);
```
why do we need this handleReflectionException method? I might prefer to use multiple catch clauses here which looks more straight forward.

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/RowSearchIterator4CountStar.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public boolean hasNext()
+    {
+        return row++ < totalRow;
```
nit: I'm not a big fan of this style though it looks concise.

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/RowSearchIterator4CountStar.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public Row next()
+    {
+        return null;
```
Is returning null an expected behavior here?

zhenxiao replied:
ping

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,215 @@
+    protected abstract void next();
+
+    /**
+     * 为了出错时好定位问题行
```
I don't think we need this comment

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,215 @@
+        try {
+            long t1 = nanoTime();
+            if (hasNext()) {
+                next(); //有可能报错
```
what happens if it throws an exception or something?

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,215 @@
+                long delta = System.nanoTime() - t1;
+
+                if (delta >= 10 * 1_000_000) {
+                    log.debug("split scan new fetch, queryId=%s currentRows=%s deltaApiCost=%sms", queryId, rows, delta / 1000_1000.0);
```
why 1000_1000.0 ?

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,215 @@
+                next(); //有可能报错
+                long delta = System.nanoTime() - t1;
+
+                if (delta >= 10 * 1_000_000) {
```
Why we have 10 times 1_000_000? should we have a constant here?

xumingming replied:
1. added a constant.
2. It is to make sure we only log every 10ms -- so it will not generate too much logs.

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreConnector.java:
```diff
@@ -0,0 +1,123 @@
+public class TablestoreConnector
+        implements Connector
+{
+    private static final Logger LOG = Logger.get(TablestoreConnector.class);
```
I might prefer using lower case 'log'

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,2003 @@
+        catch (Exception e) {
+            type = "others_error";
+            cause = e;
+            invalidate = true; //有可能客户端被关闭之类的情况
```
What happens if the client has been shutdown?  
What's 'invalidate' for? I couldn't find any usage in this class.

xumingming replied:
Indeed not needed, will delete it.

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,2003 @@
+                else {
+                    x = querySplitsFromSearchIndex(session, th, constraint, tupleDomainStr, matchedSearchIndexes);
+                    t1 = currentTimeMillis() - t1;
+                    log.info("getSplits() end, found %s indexes, query=%s splitsSize=%s apiCost=%sms.",
```
Will this generate too many logs?  

If the splits size and t1 are important can we expose them by jmx?

xumingming replied:
it only generate one log per query.

beinan reviewed (changes requested):
Sorry for my delay.  Really appreciate your contribution! Just a few coding style issues.  (I might more focus on the coding style in the first round review. ). 

This is really a big one, I will continue the review by the end of this week.  Again, thank you for the contribution!  Feel free to ping me via slack or reply the comments here if you have any questions or concern.

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreRecordCursor4RangeBlock.java:
```diff
@@ -0,0 +1,288 @@
+    @Override
+    protected void next()
+    {
+        rowIterator.next(); //有可能报错
```
remove the comment.  we can add a try/catch if you think the exception should be handled here.

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreRecordCursor4RangeBlock.java:
```diff
@@ -0,0 +1,288 @@
+                throw nonnull(name, type, Slice.class);
+            }
+
+            //上层要什么类型，底层就尝试转化成这个类型：https://work.aone.alibaba-inc.com/issue/22425902
```
comment not needed

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreRecordCursor4RangeBlock.java:
```diff
@@ -0,0 +1,288 @@
+    {
+        return callAndRecord(field, () -> {
+            String name = getColumnName(field);
+            Object b = rowIterator.getObject(name);
```
can we have a better name for 'b'?

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreRowBasedRecordCursor.java:
```diff
@@ -0,0 +1,196 @@
+    @Override
+    protected void next()
+    {
+        current = rowIterator.next(); //有可能报错
```
comment not needed

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+        switch (target.getType()) {
+            case INTEGER:
+                long v1 = target.asLong();
+                if (v1 == Long.MAX_VALUE) {
```
static import Long.MAX_VALUE

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TypeUtil.java:
```diff
@@ -0,0 +1,36 @@
+    {
+        if (value == null) {
+            String str = format("We need a nonnull instance of type[%s], but it's null.", target.getSimpleName());
+            throw new NullPointerException(str);
```
Using requireNonNull?

xumingming replied:
done

beinan commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TablestoreConstants.java:
```diff
@@ -0,0 +1,50 @@
+    static final String COLUMN_NAME_2 = "test_column2";
+    static final String COLUMN_NAME_3 = "test_column3";
+
+    static final String PK_1_int = "pk1_int";
```
All upper case for a constant

xumingming replied:
done

beinan commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestingConnectorSession.java:
```diff
@@ -0,0 +1,163 @@
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public class TestingConnectorSession
```
I think there has been a TestingConnectorSession in com.facebook.presto.testing,  do we really need to create a new one?

xumingming replied:
Removed tablestore's TestingConnectorSession and reuse presto-main's TestingConnectorSession

xumingming commented:
> Sorry for my delay. Really appreciate your contribution! Just a few coding style issues. (I might more focus on the coding style in the first round review. ).
> 
> This is really a big one, I will continue the review by the end of this week. Again, thank you for the contribution! Feel free to ping me via slack or reply the comments here if you have any questions or concern.

@beinan Thanks for the review, I will make the changes accordingly this week.

ChunxuTang reviewed (commented):
@xumingming Nice work! Just added a few comments.
BTW, could you create a doc for this new connector?

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+            String x = String.valueOf(value);
+            String str = "convert() failed for " + (toPrimaryKeyNotColumnValue ? "PrimaryKeyValue" : "ColumnValue") +
+                    ", required type is " + type + ", and type of value is " + value.getClass() + ", and value=";
+            log.error(str + x, e); //全部打印出来
```
remove the comments

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+        }
+        catch (Throwable e) {
+            String x = String.valueOf(value);
+            String str = "convert() failed for " + (toPrimaryKeyNotColumnValue ? "PrimaryKeyValue" : "ColumnValue") +
```
is it better to use String.format? your call
Also we might need a more meaningful name instead of str

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+            if (x.length() > 20) {
+                y = x.substring(0, 20) + "......";
+            }
+            throw new IllegalArgumentException(str + y, e); //打印部分
```
remove the comment

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+                acceptedTypes = new Class[] {String.class, byte[].class, Slice.class};
+            }
+        }
+        catch (Throwable e) {
```
Should we handle all the throwable here?  I think most of Errors should not handled by application code.  What do you think?

xumingming replied:
Agree with you. Changed to `Exception`

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+
+            String y = x;
+            if (x.length() > 20) {
+                y = x.substring(0, 20) + "......";
```
nit: to be consistent, is it better to use triple dot "..."?

xumingming replied:
done

beinan commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreValueUtil.java:
```diff
@@ -0,0 +1,221 @@
+                String v2 = target.asString();
+                return PrimaryKeyValue.fromString(v2 + "\0");
+            case BINARY:
+                byte[] v3 = target.asBinary();
```
I don't think v1, v2 and v3 are very meaningful name here.

xumingming replied:
I agree with you, changed it.

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreConnectorFactory.java:
```diff
@@ -0,0 +1,73 @@
+
+    public TablestoreConnectorFactory(String name, ClassLoader classLoader)
+    {
+        checkArgument(!isNullOrEmpty(name), "name is null or empty");
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,2003 @@
+    }
+
+    /**
+     * 查看总行数的查询
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,2003 @@
+            Iterator<Queue<Split>> i = ss.iterator();
+            while (i.hasNext()) {
+                Queue<Split> x = i.next();
+                Split e = x.poll(); //移除元素
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,2003 @@
+                    newList.add(e);
+                }
+                else {
+                    i.remove(); //移除队列
```
done

beinan commented:
> @xumingming Nice work! Just added a few comments.
> BTW, could you create a doc for this new connector?

Good call, doc is a new requirement to commit guideline https://lists.prestodb.io/g/presto-tsc/topic/adding_doc_requirement_to/81082141?p=,,,20,0,0,0::recentpostdate/sticky,,,20,2,0,81082141

xumingming commented:
I am back, and will make the necessary changes :)

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,215 @@
+    public long getReadTimeNanos()
+    {
+        checkClosed();
+        return (currentTimeMillis() - startTimestamp) * 1000_000;
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreConnectorFactory.java:
```diff
@@ -0,0 +1,73 @@
+            return injector.getInstance(TablestoreConnector.class);
+        }
+        catch (Exception e) {
+            throw Throwables.propagate(e);
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreRecordSetProvider.java:
```diff
@@ -0,0 +1,81 @@
+public class TablestoreRecordSetProvider
+        implements ConnectorRecordSetProvider
+{
+    private static final Logger LOG = Logger.get(TablestoreRecordSetProvider.class);
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreRecordSink.java:
```diff
@@ -0,0 +1,210 @@
+public class TablestoreRecordSink
+        implements RecordSink
+{
+    private static final Logger LOG = Logger.get(TablestoreRecordSink.class);
```
done

xumingming commented on presto-tablestore/pom.xml:
```diff
@@ -0,0 +1,238 @@
+                    </ignoredDependencies>
+                </configuration>
+            </plugin>
+		</plugins>
```
done

xumingming commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,251 @@
+    static IndexFirstSet custom(@Nonnull Set<SchemaTableName> tables)
+    {
+        if (tables.size() == 0) {
+            return none();
```
seems only be consist with cutom() method, so all place will use these methods rather than direct referencing these fields.

xumingming commented:
I have made changes according to your comments, can you help to do another round of review @beinan @ChunxuTang

xumingming commented:
> > @xumingming Nice work! Just added a few comments.
> > BTW, could you create a doc for this new connector?
> 
> Good call, doc is a new requirement to commit guideline https://lists.prestodb.io/g/presto-tsc/topic/adding_doc_requirement_to/81082141?p=,,,20,0,0,0::recentpostdate/sticky,,,20,2,0,81082141

Added the documentation.

ChunxuTang commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreConfig.java:
```diff
@@ -0,0 +1,116 @@
+    private String accessKeyId;
+    private String accessKeySecret;
+    private String instance;
+    private int clientRetrySeconds = 120;
+    private int connectTimeoutSeconds = 10;
+    private int socketTimeoutSeconds = 50;
+    private int syncClientWaitTimeoutSeconds = 60;
```
Are these settings configurable by users? If yes, could you add documentation of these settings in the alibaba cloud tablestore doc?

xumingming replied:
yes, they are configurable, but I'd prefer merge this first PR into prestodb first, and later I can expose more settings for user to tune.

ChunxuTang commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,1994 @@
+                .indexName(indexName)
+                .scanQuery(ScanQuery.newBuilder()
+                        .query(buildQueryForCurrentConditions(tupleDomain))
+                        .limit(2000)
```
Just curious why the value here is 2000. Is it from the empirical experience?

xumingming replied:
It is recommended by the tablestore team, it is not limiting the result count to 2000, it is limiting to only return 2000 for every request.

ChunxuTang commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreMetadata.java:
```diff
@@ -0,0 +1,152 @@
+        List<SchemaTableName> list = listTables(session, prefix.getSchemaName());
+        for (SchemaTableName stn : list) {
+            String qtn = prefix.getTableName();
+            if (qtn != null && !qtn.equalsIgnoreCase(stn.getTableName())) { //如果tableName==null，就匹配全部；否则按tableName过滤
```
Could you remove the comment or translate it into English?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestorePlugin.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static ClassLoader getClassLoader()
+    {
+        ClassLoader tcl = Thread.currentThread().getContextClassLoader(); //PrestoServer对应启动线程的类加载器
```
Could you remove the comment?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreTransactionHandle.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    public String toString()
+    {
+        return "TablestoreTransactionHandle{" +
+                "uuid=" + uuid +
+                '}';
```
Looks that this is not that wide to split into 3 lines. But either way works for me. Your call.

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreFacade.java:
```diff
@@ -0,0 +1,1105 @@
+        d.clear();
+        d.put(col3, Domain.onlyNull(INTEGER));
+        f = transformToFilterV2(d);
+        doFilterAndEquals("c3 IS NULL", f); //非主键
```
Could you remove the comment here?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreFacade.java:
```diff
@@ -0,0 +1,1105 @@
+        d.clear();
+        d.put(pk0, Domain.onlyNull(INTEGER));
+        f = transformToFilterV2(d);
+        doFilterAndEquals("", f); //主键
```
Could you remove the comment here?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreFacade.java:
```diff
@@ -0,0 +1,1105 @@
+    }
+
+    /**
+     * 类似于：0:(min,0) 1:(0,1) 2:(1,2) 3:(2,3) 4:(3,4) 5:(4,5) 6:(5,6) 7:(6,7) 8:(7,8) 9:(8,max)，第一个是分区，第二个是min，第三个是max
```
If the comment here is necessary, could you translate it into English?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreFacade.java:
```diff
@@ -0,0 +1,1105 @@
+        List<Split> rList = divideAndMergeAndShuffle(4, list);
+        assertEquals("0:(min,3) 1:(3,5) 2:(5,7) 2:(7,max)", print(rList));
+
+        //希望合并到3份，但也只能合并到4份，因为有些合并不了了
```
Maybe remove the comment here? I guess no matter we merge them to 3 or 4, it won't influence the major features.

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreFacade.java:
```diff
@@ -0,0 +1,1105 @@
+            long n2 = System.currentTimeMillis();
+
+            assertEquals(10000, rList.size());
+            System.out.println("测试性能-" + i + ": " + (n2 - n1) + "ms");
```
Maybe change the words to English?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreSessionProperties.java:
```diff
@@ -0,0 +1,370 @@
+
+        ConnectorSession cs = session();
+        int x = getQueryVersion(cs);
+        assertEquals(2, x); //默认不开启
```
Maybe remove the comment?

xumingming replied:
done

ChunxuTang commented on presto-tablestore/src/test/java/com/facebook/presto/tablestore/TestTablestoreSessionProperties.java:
```diff
@@ -0,0 +1,370 @@
+
+        TestingConnectorSession cs = session();
+        boolean x = enablePartitionPruning(cs);
+        assertTrue(x); //默认开启
```
Same as above.

xumingming replied:
done

ChunxuTang reviewed (commented):
@xumingming Thanks for your huge work! Generally looks good to me. Just left some minor suggestions in the code. @beinan Could you also have another round of review?

BTW, I noticed that there are quite a number of commits in this PR and most of them are just for fixing comments or updating versions. @xumingming When you complete changes, could you merge them into one commit? Thanks~

xumingming commented:
@ChunxuTang updated the code according to all the comments, and merged all the relative commits into one commit.

beinan reviewed (approved):
@ChunxuTang @zhenxiao do you two wanna take a second look?  otherwise I will merge this PR

ChunxuTang reviewed (approved):
LGTM!

xumingming commented:
Let's merge it?

zhenxiao commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+``tablestore.endpoint``            The endpoint of tablestore.
+``tablestore.access-key-id``       The AccessKeyId used to access Tablestore.
+``tablestore.access-key-secret``   The AccessKeySecret used to access Tablestore.
+``tablestore.instance=demo``       The name of the instance we'd like to connect to.
```
remove `=demo`

xumingming replied:
done

zhenxiao commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+
+    connector.name=tablestore
+    tablestore.endpoint=https://demo.cn-hangzhou.ots.aliyuncs.com
+    tablestore.access-key-id=demo
```
is `demo` intentionally used here? or, we could use `id`, `key`, `instance` or other names

xumingming replied:
done

zhenxiao commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+The endpoint of tablestore, e.g. https://demo.cn-hangzhou.ots.aliyuncs.com.
+
+``tablestore.access-key-id``
+^^^^^^^^^^^^^^^^^^^^
```
alignment for `^`

xumingming replied:
done

zhenxiao commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+The AccessKeyId used to access Tablestore.
+
+``tablestore.access-key-secret``
+^^^^^^^^^^^^^^^^^^^^^
```
alignment

zhenxiao commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+
+
+``tablestore.instance``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
same, alignment

xumingming replied:
done

zhenxiao commented on presto-main/etc/catalog/tablestore.properties:
```diff
@@ -0,0 +1,12 @@
+
+connector.name=tablestore
+tablestore.endpoint=https://demo.cn-hangzhou.ots.aliyuncs.com
+tablestore.access-key-id=demo
```
again, is `demo` intentionally used here? or, we might try other names

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/CustomRetryStrategy.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.concurrent.TimeUnit;
+
+public class CustomRetryStrategy
```
how about timeoutRetryStrategy?

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/CustomRetryStrategy.java:
```diff
@@ -0,0 +1,43 @@
+
+    public CustomRetryStrategy(int timeoutSeconds)
+    {
+        super(timeoutSeconds, TimeUnit.SECONDS);
```
static import SECONDS

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+
+public class IndexFirstSet
+{
+    private static final String INDEX_HINT_PARSE_ERROR_MESSAGE = "use hint like these: " +
```
use hints like:

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+    private static final String INDEX_HINT_PARSE_ERROR_MESSAGE = "use hint like these: " +
+            "1)use index if possible -> 'tablestore-index-first=auto' " +
+            "2)do not use any index, default -> 'tablestore-index-first=none' " +
+            "3)use indexes of tables that specified -> 'tablestore-index-first=[xxDb.yyTable, zzTable, ...]' " +
```
xxDb.yyTable, zzTable -> db1.table1, db2.table2

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+        this.type = type;
+    }
+
+    /**
```
do we need this comment?

xumingming replied:
removed

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+        return NONE;
+    }
+
+    /**
```
is this comment useful?

xumingming replied:
removed

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+            return none();
+        }
+        else {
+            IndexFirstSet x = new IndexFirstSet(Type.CUSTOM);
```
s/x/indexFirstSet/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+        if (!unit.contains(".")) { //'schema'.'table'
+            s = currentConnectionSchema.<IllegalArgumentException>orElseThrow(() -> {
+                String str = "Can't obtain the schema of the table[" + unit + "] from current connection for hint '"
+                        + hintKey + "'";
```
merge into the line above

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java:
```diff
@@ -0,0 +1,264 @@
+import static com.facebook.presto.tablestore.TablestoreSessionProperties.MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS;
+import static com.google.common.base.Preconditions.checkState;
+
+public class IndexFirstSet
```
could we have a better name? IndexFirstSet is king of misleading

xumingming replied:
renamed to IndexSelectionStrategy

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/RowSearchIterator4CountStar.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.Iterator;
+
+public class RowSearchIterator4CountStar
```
do you mean RowSearchIteratorForCountStar? how about rename it as: CountStarRowIterator?

xumingming replied:
nice name, changed.

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,214 @@
+
+    protected boolean closed;
+    protected int rows;
+    protected long apiNano;
```
do you mean `apiInNanos`?

xumingming replied:
yeah, changed it.

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,214 @@
+    protected long apiNano;
+    protected long readNano;
+
+    public TablestoreAbstractRecordCursor(TablestoreFacade tablestoreFacade, ConnectorSession session, TablestoreSplit split,
```
all parameters in one line, or each in a separate line

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,214 @@
+        return columnHandles[field].getColumnType();
+    }
+
+    protected String getColumnName(int idx)
```
s/idx/index/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,214 @@
+
+    protected abstract void next();
+
+    protected abstract String currentPrimaryKeyAsDebug();
```
shall we have a better name? hard to understand `currentPrimaryKeyAsDebug`

xumingming replied:
It is used to print the current primary key(type and value), mainly for debug purpose when there is something wrong, renamed to currentPrimaryKeyInfo()

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,214 @@
+
+    protected abstract String currentPrimaryKeyAsDebug();
+
+    protected String toString4Value(Object value)
```
s/toString4Value/toString/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java:
```diff
@@ -0,0 +1,214 @@
+            return "null";
+        }
+        else if (value instanceof byte[]) {
+            String x = Base64.getEncoder().encodeToString((byte[]) value);
```
please not use `x`, `y` as variable name, maybe encodeValue?

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreClientHelper.java:
```diff
@@ -0,0 +1,107 @@
+
+    protected static SyncClient createSyncClient(TablestoreConfig config, Optional<ExecutorService> callback)
+    {
+        ClientConfiguration cfg = new ClientConfiguration();
```
s/cfg/configuration/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreClientHelper.java:
```diff
@@ -0,0 +1,107 @@
+        cfg.setSocketTimeoutInMillisecond(config.getSocketTimeoutSeconds() * 1000);
+        cfg.setSyncClientWaitFutureTimeoutInMillis(config.getSyncClientWaitTimeoutSeconds() * 1000);
+
+        SyncClient cli = new SyncClient(config.getEndpoint(), config.getAccessKeyId(), config.getAccessKeySecret(), config.getInstance(), cfg);
```
s/cli/client/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreCountStarCursor4SearchIndex.java:
```diff
@@ -0,0 +1,107 @@
+
+import static java.lang.System.currentTimeMillis;
+
+public class TablestoreCountStarCursor4SearchIndex
```
s/TablestoreCountStarCursor4SearchIndex/SearchIndexTableStoreCursor/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,1994 @@
+            type = "others_error";
+            cause = e;
+        }
+        String err = "Error communicate with TableStore. Detailed message: %s() failed, type=%s, traceId=%s, requestId=%s, params={%s}, because: %s";
```
s/err/error/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,1994 @@
+            cause = e;
+        }
+        String err = "Error communicate with TableStore. Detailed message: %s() failed, type=%s, traceId=%s, requestId=%s, params={%s}, because: %s";
+        String pStr = Joiner.on(",").join(params);
```
let's not use `pStr` as variable name

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,1994 @@
+                .collect(toList());
+    }
+
+    public List<? extends ConnectorSplit> getSplits(ConnectorSession session, TablestoreTableLayoutHandle ol)
```
s/ol/layoutHandle/g

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,1994 @@
+    {
+        ListTableResponse listTableResponse = syncClient.listTable();
+        return listTableResponse.getTableNames().stream()
+                .map(x -> new SchemaTableName(config.getInstance(), x))
```
let's not use `x` as variable name

xumingming replied:
done

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java:
```diff
@@ -0,0 +1,1994 @@
+        TypeSignature ts = TypeSignature.parseTypeSignature(columnType.toLowerCase(Locale.getDefault()));
+        switch (ts.getBase()) {
+            case "boolean":
+                return BooleanType.BOOLEAN;
```
static import BOOLEAN, INTEGER, DOUBLE, VARCHAR, VARBINARY

xumingming replied:
done

zhenxiao reviewed (changes requested):
nice work, @xumingming 
some minor coding style issues

xumingming commented:
@zhenxiao For the `return null` in `RowSearchIterator4CountStar.java`, it can return null because the returned value will not be used, it is an optimization for `count(*)`.

xumingming commented:
@zhenxiao Any further comments?

xumingming commented:
@zhenxiao Any further comments?

xumingming commented:
@zhenxiao Any further comments?

xumingming commented:
@beinan @zhenxiao @ChunxuTang Any further comments?

xumingming commented:
@zhenxiao Any further comments?

xumingming commented:
@zhenxiao Any comments?

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    private final Type type;
+    private int maxPercent = -1;
```
s/maxPercent/maxPercentage/g

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        this.type = type;
+    }
+
+    static IndexSelectionStrategy none()
```
do we need none() and auto()? could we use NONE and AUTO directly?

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        throw new IllegalStateException("Can't get max rows for '" + type.getCode() + "' type");
+    }
+
+    /**
```
let's remove this comment. seems not useful

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        return tables.contains(table);
+    }
+
+    public String backToString()
```
backToString is kind of misleading, shall we use toString()?

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static IndexSelectionStrategy parse(Optional<String> currentConnectionSchema, String hintKey, String hintValue)
+    {
+        if (StringUtils.isBlank(hintValue)) {
```
static import StringUtils.isBlank

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        if (StringUtils.isBlank(hintValue)) {
+            return IndexSelectionStrategy.none();
+        }
+        String hint = hintValue.trim().toLowerCase(Locale.ENGLISH);
```
static import Locale.ENGLISH

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        }
+        String pattern = "^" + Type.THRESHOLD.getCode() + ":\\s*(\\d+)\\s*(%)?$";
+        if (hint.matches(pattern)) {
+            String err = "Invalid 'threshold' hint value '" + hintValue + "' of hint key '" + hintKey + "', regular pattern='" + pattern + "'";
```
s/err/error/g

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        String pattern = "^" + Type.THRESHOLD.getCode() + ":\\s*(\\d+)\\s*(%)?$";
+        if (hint.matches(pattern)) {
+            String err = "Invalid 'threshold' hint value '" + hintValue + "' of hint key '" + hintKey + "', regular pattern='" + pattern + "'";
+            Matcher m = Pattern.compile(pattern).matcher(hint);
```
s/m/matcher/g

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+
+            if (percent != null) {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT) {
+                    err = "Invalid 'threshold:${maxPercent}%' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
```
shall we inline error message creation in IllegalArgumentException? instead of having an err variable?

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+            }
+            else {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS) {
+                    err = "Invalid 'threshold:${maxRows}' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
```
please simplify the error message

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+            if (StringUtils.isBlank(hint)) {
+                return IndexSelectionStrategy.none();
+            }
+            String[] ss = hint.split("[ ;,]+");
```
replace ss to something meaningful

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+                    .filter(StringUtils::isNotBlank)
+                    .map(String::trim)
+                    .map(String::toLowerCase)
+                    .map(x -> checkAndAssemblySchemaTableName(currentConnectionSchema, hintKey, x))
```
s/x/hint/g

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+            String hintKey,
+            String unit)
+    {
+        String s;
```
replace s with something meaningful

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java:
```diff
@@ -0,0 +1,257 @@
+        String s;
+        if (!unit.contains(".")) { //'schema'.'table'
+            s = currentConnectionSchema.<IllegalArgumentException>orElseThrow(() -> {
+                String str = "Can't obtain the schema of the table[" + unit + "] from current connection for hint '" + hintKey + "'";
```
how about inline error message `Can't obtain the schema of`, instead of creating an `str` variable
also, please do not use `str` as variable name, let's try something meaningful

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,65 @@
+        try {
+            return field.get(target);
+        }
+        catch (IllegalAccessException ex) {
```
s/ex/exception/g

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,65 @@
+        }
+        catch (IllegalAccessException ex) {
+            throw new IllegalStateException("Unexpected reflection exception - " + ex.getClass().getName() + ": "
+                    + ex.getMessage());
```
ex.getMessage() added to line above

zhenxiao commented on presto-tablestore/src/main/java/com/facebook/presto/tablestore/ReflectionUtils.java:
```diff
@@ -0,0 +1,65 @@
+    public static void makeAccessible(Field field)
+    {
+        if ((!Modifier.isPublic(field.getModifiers()) || !Modifier.isPublic(field.getDeclaringClass().getModifiers()) || Modifier
+                .isFinal(field.getModifiers())) && !field.isAccessible()) {
```
let's put .isFinal the same line with Modifier. either break the lines by`||`, or put everything in the same line

zhenxiao reviewed (changes requested):
thanks, @xumingming 
a few more comments

xumingming commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+``tablestore.endpoint``            The endpoint of tablestore.
+``tablestore.access-key-id``       The AccessKeyId used to access Tablestore.
+``tablestore.access-key-secret``   The AccessKeySecret used to access Tablestore.
+``tablestore.instance``            The name of the instance we'd like to connect to.
```
not hostname, it is an instance, I think TableStore user can understand it.

xumingming commented on presto-docs/src/main/sphinx/connector/tablestore.rst:
```diff
@@ -0,0 +1,74 @@
+``tablestore.instance``
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The name of the instance we'd like to connect to.
```
I think it does not need to change.

xumingming commented:
@zhenxiao all comments resolved, can you help review again?
